formats/hocon/src/test/kotlin/kotlinx/serialization/hocon/HoconRootObjectsTest.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2017-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.hocon
+
+import com.typesafe.config.ConfigFactory
+import kotlinx.serialization.Serializable
+import org.junit.Assert.*
+import org.junit.Ignore
+import org.junit.Test
+
+class HoconRootMapTest {
+    private val configForRootMap = """
+        key1 {
+            a = "text1"
+            b = 11
+            }
+        
+        key2 {
+            a = "text2"
+            b = 12
+        }
+        
+        key3 {
+            a = "text3"
+            b = 13
+        }
+    """
+    private val configWithEmptyObject = "{}"
+    private val configWithRootList = "[foo, bar]"
+    private val emptyConfig = ""
+
+    @Serializable
+    data class CompositeValue(
+            val a: String,
+            val b: Int
+    )
+
+    @Test
+    fun testConfigWithRootMap() {
+        val config = ConfigFactory.parseString(configForRootMap)
+        val obj = Hocon.decodeFromConfig<Map<String, CompositeValue>>(config)
+
+        assertEquals(CompositeValue("text1", 11), obj["key1"])
+        assertEquals(CompositeValue("text2", 12), obj["key2"])
+        assertEquals(CompositeValue("text3", 13), obj["key3"])
+    }
+
+    @Test
+    fun testEmptyObjectDecode() {
+        val config = ConfigFactory.parseString(configWithEmptyObject)
+        //non-null map decoded from empty object as empty map
+        val map = Hocon.decodeFromConfig<Map<String, String>>(config)
+        assertTrue(map.isEmpty())
+
+        //nullable map decoded from empty object as null - not obvious
```

Comment:
I think this behavior is actually ok: given that `Hocon.decodeFromConfig<ComplexValue?>(emptyConfig)` returns null (or I really hope it does so, needs testing), this behavior is more consisntent that it may looks like at first sight.

Anyway, we can change it if we really want. Since original `config` is Java library, it seems that it cannot distinguish between nullable and non-null types, so this part is likely caused by our parser.

Reply:
Disagree. In case of static-size types (e.q. some custom ComplexValue) it's better to return null.

But as user with variable-size structure (Map) I expect size of map will be equals with JSON object properties count. In config object exists, so it's surprise for me - function behavior changed after changing return type to nullable, not changing of config string.
For example Jackson does as I said
objectMapper.readValue("{}", Map::class.java)  returns empty map and it is expected - I see begin object symbol and end object symbol - so object exists but not contains any properties.

Returning null is not a bug but some users can be confused this behavior.

Reply:
I've tested json and
```kotlin
        println(Json.decodeFromString<Map<String, String>>(emptyJson))
        println(Json.decodeFromString<Map<String, String>?>(emptyJson))
```

both give empty map as a result. And decoding even nullable user type from "{}" gives an error.

So we'd indeed better change that, if it's possible

Reply:
Change HOCON decoder behavior? In this PR?

Reply:
Yes

Reply:
For this case (nullable map)